src/Microsoft.NetCore.Analyzers/Core/Security/UseAutoValidateAntiforgeryToken.cs

```diff
@@ -0,0 +1,361 @@
+                        if (derivedControllerTypeSymbol.GetAttributes().Any(s => s_AntiForgeryAttributeRegex.IsMatch(s.AttributeClass.Name)))
+                        {
+                            usingValidateAntiForgeryAttribute = true;
+                            currentControllerIsSafe = true;
+                        }
+
+                        foreach (var actionMethodSymbol in derivedControllerTypeSymbol.GetMembers().OfType<IMethodSymbol>())
+                        {
+                            if (actionMethodSymbol.MethodKind == MethodKind.Constructor)
+                            {
+                                continue;
+                            }
+
+                            if (actionMethodSymbol.IsPublic() &&
+                                !actionMethodSymbol.IsStatic &&
+                                !actionMethodSymbol.HasAttribute(nonActionAttributeTypeSymbol))
+                            {
+                                // The method is protected by a validate anti forgery token attribute
+                                if (currentControllerIsSafe || actionMethodSymbol.GetAttributes().Any(s => s_AntiForgeryAttributeRegex.IsMatch(s.AttributeClass.Name)))
+                                {
+                                    usingValidateAntiForgeryAttribute = true;
+
+                                    if (!actionMethodSymbol.GetAttributes().Any(s => s.AttributeClass.GetBaseTypes().Contains(httpMethodAttributeTypeSymbol)))
+                                    {
+                                        actionMethodNeedAddingHttpVerbAttributeSymbols.Add(actionMethodSymbol);
+                                    }
+                                }
+                                else
+                                {
+                                    var httpVerbAttributeTypeSymbol = actionMethodSymbol.GetAttributes().FirstOrDefault(s => httpVerbAttributeTypeSymbols.Contains(s.AttributeClass));
+
+                                    if (httpVerbAttributeTypeSymbol != null)
+                                    {
+                                        var attributeName = httpVerbAttributeTypeSymbol.AttributeClass.Name;
+                                        actionMethodSymbols.Add(
+                                            (actionMethodSymbol,
+                                            attributeName.EndsWith("Attribute", StringComparison.Ordinal) ? attributeName.Remove(attributeName.Length - "Attribute".Length) : attributeName));
+                                    }
+                                    else if (!actionMethodSymbol.GetAttributes().Any(s => s.AttributeClass.GetBaseTypes().Contains(httpMethodAttributeTypeSymbol)))
+                                    {
+                                        actionMethodSymbols.Add((actionMethodSymbol, "Http"));
+                                    }
+
```

Comment:
What we want is:
- If the controller has an antiforgerytoken attribute on it, no need for any diagnostics.
- Else, if the method has an antiforgery token attribute on it, no need for any diagnostics.
- Else, if the method has any HttpPost/HttpDelete/HttpPatch/HttpPut attribute on it, need a diagnostic.  Note that if the method only has an HttpGet attribute (or other "safe read" HTTP method) on it, then no need for a diagnostic in that case.
- Else, if the method has no Http attribute on it whatsoever, then diagnostic for the missing HttpMethod attribute.